draft: Logging on Devices (#409)
memfault/interrupt: _drafts/device-logging.md

---
title: Trade-offs and considerations for logging on embedded devices
description: "Trade-offs and considerations for logging on embedded hardware"

> author: victorlai

Reviewer: semi-related, might be good to link to the logging section from my Android debugging blog post.
https://interrupt.memfault.com/blog/debugging-android-devices#logging

It has a good section calling out `lnav`. Not sure if that goes against the second half of the post discussing how to get away from manually reading logs but its still a handy tool.

> author: victorlai

Reviewer 2: Linking back to your logging section could provide some continuity of content -- mention that we've talked about this before and we're building upon it. I've seen this on the blog before and liked it. @Blisse what do you think? Something like:

"In [a previous article on Android debugging](https://interrupt.memfault.com/blog/debugging-android-devices#logging), we mentioned different sets of logs on Android devices. Without good logging practices, these logs can be difficult to use."

> author: victorlai

Author: I like the idea of referencing a previous article, but I'm having trouble establishing continuity between that article and this one. The intro here is specifically non-Android focused, so we'd have to bring in Android maybe in the excerpt. Best bet would probably be somewhere below, probably in the section about Log Cache Size when we talk about the different ways that logs can be stored.

> author: victorlai

Reviewer 2: Fair point. You could toss it in your Additional Reading section. I'm not passionate about it though, so will leave to you and @phoenixuprising

> author: victorlai

Reviewer: yeah, I agree wigh @gminn here. Additional reading sounds fine to me.

---

Logging on client devices comes with a host of interesting, unique challenges that aren't found on the server &mdash; devices are generally more memory, bandwidth, space, and CPU constrained than a server, and devices often take weeks, months, or years to update, rather than days or hours. This has interesting ramifications on how we log: we can't log as much data as we'd want to, logs have a more finite lifespan, and the format of logs are much more permanent, so if we don't log the right data in the right way to begin with, it can be months before we ship the next release with new logs. All that being said, there's also a lot of literature and learnings from logging on the server that we can also apply on the client side.

In this article, I discuss some of the problems I’ve encountered in both producing and analyzing device logs over the years, on single digit number of devices up to hundred of thousands, and concepts I’ve applied to manage that complexity. My background is Android, so some of the examples may skew towards much more powerful hardware than simple microcontrollers.

{% include newsletter.html %}

{% include toc.html %}

> ## Log collection

Author: I like it! (though I wish the sub sub headings were in the TOC too lol)

> ## Log collection

Reviewer 2: I think sub sub headings might be clutter the TOC a bit from this snapshot view but it would technically be more complete. Probably more important for documentation to have a complete TOC than articles since we don't typically need good nav through an article

> 

Reviewer 2: A quick sentence introducing this section will help with flow

I've been lucky in that embedded Android devices are generally not very disk space constrained, so I’ve had the benefit of having 10's of MBs of logs to store (and comb) through. But even so, that would only constitute maybe 2-3 hours of logs, depending on how busy the device was, and there was a constant trade-off with how much information was being logged.

### On-location versus Remote collection

One of my favorite often overlooked questions once you have a product device is, "how do I get the logs off of my device and onto my laptop"?

During development, this is as simple as hooking a cable directly to the device and downloading the logs files off the filesystem directly. But when the device is in the field, it’s not so simple.

I distinctly remember going to out to a customer’s store and connecting to their network so I could pull logs off of a remotely accessible device, just to diagnose an issue they were having.

> This is of course, unsustainable, if you have hardware that’s distributed across multiple locations, and if you have the want to collect logs faster than just waiting until you are physically connected to the device again.

Author: Thanks, "physical access" is the key phrase I could not recall lol

The first alternative that comes to mind is a push system for retrieving logs &mdash; send a notification from the server to the device, and the device will upload the logs when it receives the message. This can take the form of an actual push notification mechanism with a connected web socket, or something as simple as having the device perform a poll every once in a while and collecting the log if the poll tells it to.

The problem with using a push to upload logs is that by the time you send the push, or by the time that the device receives it, the log cache might not contain the data from the time of the interest anymore. At a previous company, it was often the case that because our log cache often saved the last 2-3 hours of logs, by the time our engineers were informed about an incident in the field, a day would have passed and the logs that we’d retrieve wouldn’t have any useful information from the time of the incident.

> One of the cooler [log collection triggers](https://docs.memfault.com/docs/android/android-logging#log-collection) we developed here at Memfault is what I like to call "Just in time" log collection. Instead of having the user of the device click a button or someone send a push to the device, we can automatically detect if an abnormal event happened, and then mark the log files around that time as significant, where they will then be uploaded at the next convenient moment. This allows us to capture the exact log files that are relevant to an incident, if we can automatically detect that abnormal event.

Reviewer 2: I feel like this docs link doesn't show off cool log collection triggers the way I'd want it to...is there somewhere else we can link to the illustrates this, or maybe we can beef up this section of the docs before publishing? Would super love to see this link wow people if they click on it!

Author: Haha I had this same feeling when I wrote this - let me make a quick update to the docs. I thought we had something more substantial to be honest.

Author: https://github.com/memfault/memfault-docs/pull/755

Reviewer 2: Sweet!! Love it

Reviewer 2: Going to keep this thread unresolved until you merge it so we don't forget

<p align="center">
  <img width="850" src="{% img_url blog-device-logging/jit-logs.png %}" alt="Just in time log collection" />
</p>

## Logging practices

> 

Reviewer 2: A quick sentence introducing this section will help with the flow

### Write logs for the target audience

It’s important to consider the target audience of your logs, especially as the use of your device grow over the years, and more people and more teams are involved in its development.

When you’re only logging for yourself — really, anything goes, as you’ll have all the context to make interpreting the logs easier (though you might still want to give your future self an easier time too). But when your logs might be used by your teammates, or even other teams with varying context and skills, you might want to use some strategies to make sure that the logs are also effective for them.

#### Make logs actionable

> During development, we can be guilty of logging in a way that’s technically accurate rather than actionable, because we assume the eventual reader of the log will have as much context and knowledge of the domain as we do, and they’ll be able to diagnose the remedy themselves.

Reviewer 2: Really liked how you made this personal with a description of your experience with it. Nice!

```kotlin
when (val name = reader.nextName()) {
 VERSION -> Logger.w("Invalid key: '$name'")
  else -> Logger.w("Unhandled key: '$name'")
}
```

Instead of only adding the technical reason for the failure, I’ve found it helpful to add information about how the error is expected to be resolved, in addition to the domain specific information about what went wrong [^meaningfulmessages].

```kotlin
when (val name = reader.nextName()) {
 VERSION -> Logger.w(
    "Invalid key: '$name', the key should only be written, not read"
  )
  else -> Logger.w(
  "Unhandled key: '$name', is the parser version behind the writer version?"
 )
}
```

Adding additional context can jumpstart the debugging effort, as now the person debugging has an avenue to investigate immediately after seeing the log.

#### Link logs to additional documentation

It doesn't feel right to write a paragraph in the log, but it can often take a few sentences to actually explain what the error being logged is.

```kotlin
try {

>   ...

Reviewer 2: The dots render a little funky, so add it add a comment:
```suggestion
  // ...
```

} catch (e: RemoteException) {
  Logger.w("Unable to connect to Reporter, this happens if Reporter is so old that it does not contain the Service at all.", e)
}
```

Instead of logging a large paragraph, we can log a reference link to somewhere else where the paragraph is written. This is similar to how SDKs will also contain a glossary of all the exceptions it may throw, what each error means, and documentation on how to resolve each one. We can apply that same logic with our own logging.

```kotlin
try {

>   ...

Reviewer 2: Same here:
```suggestion
  // ...
```

} catch (e: RemoteException) {

>   // This can happen if Reporter is so old that it does not contain the Service at all

Reviewer 2: Super nit: How about removing this comment? A few reasons:
1. I see the use of the reference links not only making logs more readable, but also making the source (comments + code!) more concise. Just shifting the contents of the log message into a comment above defeats that purpose.
2. We are suggesting that a glossary of terms should be created, which will define `REPORTER_SERVICE_NOT_FOUND`.  In that case, I think it is better to have one source of truth and avoid extra comments like this.

Author: Yeah I tried to have kind of 3 examples here to show moving from a long log to a comment to additional docs but maybe a bit unnecessary.

 
```markdown
V bort    : Payment started: payment_session_id=0xFF
V bort    : Awaiting response: payment_session_id=0xFF
V bort    : Payment rejected 500: payment_session_id=0xFF, msg=...
```

#### Unique logs

A major impedance to reading logs is when the exact same string is re-used multiple times.

```markdown
V bort    : Payment started: payment_session_id=0xFF
```

```kotlin
fun startPayment() {
    Logger.v("Payment started: payment_session_id=$id")
    ...
}

fun retryPayment() {
    Logger.v("Payment started: payment_session_id=$id")
    ...
}
```

If you see this log and trying to diagnose the underlying issue, you might have no clue which line of code it’s actually referring to, so it’s important that all our logs are unique from one another.

### Log for computers first

We'll never really stop reading logs, but it can become a chore to read 1000s of lines of logs, or look to see if multiple sets of logs from multiple devices may be experiencing the same issue. One of the best ways to get out of the business of reading logs, is to have our computers read it for us, and display it in a more digestible format.

> At past companies, we developed command line tools that would read log files and print out a text summary of all the exceptions, any network connectivity issues detected, and various other useful information. The benefit of this was massive — the tool could analyze logs much faster than we could individually and in much less time, we could share the tool with our colleagues and save time across the board, we could iterate on adding new features to the tool as the team grew and new scenarios were discovered.

Reviewer 2: A suggestion for clarity:
```suggestion
At my past companies, we developed a shared command line tools that would read log files and print out a text summary of all the exceptions, any network connectivity issues detected, and various other useful information. The benefit of this was massive; not only did the tool analyze logs much faster than we could individually, but having a shared solution meant we all cashed out on this time savings, we all agreed upon the interpretation of different log messages, and we could add new features to the tool as the team grew and new scenarios were discovered.
```

We can take another page out of server logging best practices [^structuredlogging], and use structured logging. Logs by default are unstructured strings. If we logged more often in a structured format that was easily machine parsable, we could greatly simplify parsing and organizing the data. This is especially useful when we want to eventually parse specific information in the log, otherwise we'd have to write a Regex to parse the log.

```markdown
E bort    : {"Network error": "-55", "RSSI": 61, "SSID": "Home-Guest"}
```

JSON is the most common format for structured logs, and can be written manually in code fairly quickly. It's also fairly readable compared to the unstructured log, too.

#### Skip the reading and build visualizations

The end state of logging is ideally to never read the raw logs ourselves. Once we can empower computers to read all sorts of logs, we can build the tools to parse and then visualize the information, and save ourselves the effort of piecing together all the information in our head.

<p align="center">
  <img width="850" src="{% img_url blog-device-logging/logging-fidelity.png %}" alt="Wi-Fi connection status over time" />
</p>

> If this sounds familiar, this is pitch for Memfault! Instead of reading logs, we can acquire a lot of the same data that the log would have produced, and then display it on a browsable timeline in extremely high resolution.

Reviewer 2: Could you add some more details on how we got to this image? Are these metrics? 😉

## Conclusion

Logging is a fundamental part of our day to day debugging toolset, and it's hard to imagine it ever going away entirely. But we all agree, there is a lot of noise in our logs, there are really no standards for logging, and there's a lot we could probably improve how we log, if we only had the time. Hopefully I've shared some new or helpful ideas around logging that you can put to use today!

<!-- Interrupt Keep START -->
{% include newsletter.html %}

{% include submit-pr.html %}
<!-- Interrupt Keep END -->

{:.no_toc}

## Additional Reading

* [Logging v. instrumentation](https://peter.bourgon.org/blog/2016/02/07/logging-v-instrumentation.html)
* [Tracing: structured logging, but better in every way](https://andydote.co.uk/2023/09/19/tracing-is-better/)

> ## References

Reviewer 2: So References don't show up in the TOC:

```suggestion
{:.no_toc}

## References
```

---

> title: Trade-offs and considerations for logging on embedded devices

Reviewer: Title doesn't seem consistent with the rest of our articles.

Author: years of square has made me prefer sentence casing to title casing 🙈

description: "Trade-offs and considerations for logging on embedded hardware"
author: victorlai
tags: [android, debugging, logging]
---

> Logging on embedded devices comes with a host of interesting, unique challenges that aren't found on the server &mdash; devices are generally more memory, bandwidth, space, and CPU constrained than a server, and devices often take weeks, months, or years to update, rather than days or hours. This has interesting ramifications on how we log: we can't log as much data as we'd want to, logs have a more finite lifespan, and the format of logs are much more permanent, so if we don't log the right data in the right way to begin with, it can be months before we ship the next release with new logs. All that being said, there is still a lot of literature and learnings from logging on the server that we can also apply on the client side.

Reviewer: nit: the use of interesting, unique challenges reads weird to me.
```suggestion
Logging on embedded devices comes with a host of interesting and unique challenges that aren't found on the server &mdash; devices are generally more memory, bandwidth, space, and CPU constrained than a server, and devices often take weeks, months, or years to update, rather than days or hours. This has interesting ramifications on how we log: we can't log as much data as we'd want to, logs have a more finite lifespan, and the format of logs are much more permanent, so if we don't log the right data in the right way to begin with, it can be months before we ship the next release with new logs. All that being said, there is still a lot of literature and learnings from logging on the server that we can also apply on the client side.
```

Author: Yeah it's intentionally more stream of conscious writing than formal essay. But and works a bit better anyways.

The amount of disk space chosen has a large influence on logging &mdash; it affects how verbose your logs can be, how long your logs are be retained, what information you'll want prioritize logging, how you'll store the logs, and more. You’ll want to choose the maximum amount of space you can afford, where the logs you get back contain enough useful information at the timeframe of the incident. But you'll also want to change how you log to accomodate for the amount of space, such as using an abbreviated short form instead of full words.

I've been lucky in that embedded Android devices are generally not very disk space constrained, so I’ve had the benefit of having 10's of MBs of logs to store (and comb) through. But even so, that would only constitute maybe 2-3 hours of logs, depending on how busy the device was, and there was a constant trade-off with how much information was being logged.

> ### On-location v. Remote Collection

Reviewer: TIL `v.` is used in legal docs to shorten versus... and I don't like it 🙃. This would be personal preference I prefer `vs.`.

Reviewer 2: Fair point, now that you've mentioned it, I can't unsee the legal usage. `vs.` works for me

> One of my favorite often overlooked questions when you develop a new embedded device product is, "how do I get the logs off of my production device and onto my computer"? During development, this could be as easy as connecting a cable directly to the device and downloading the logs files off the filesystem directly. But when the device is in the field, it’s not so simple.

Reviewer: We jump from Log Collection Cache Size directly to Log Collection. Is it worth adding a section talking about the pros and cons of in-memory vs writing to disk? Android defaults to in-memory. MCU (from my very limited experience) will often times *just* stream them over serial. The main things to balance that I can think of currently are memory constraints vs disk wear.

Author: I'm not sure in memory vs streaming makes sense in the context of this article, because if it's not persisted to the filesystem, then there's no real way of remote debugging it? I also kind of prefaced that the article has an Android slant.

But I do like that it adds some more weight to the kind of bare section.

Reviewer: lgtm

> I distinctly remember going to out to a customer’s store and connecting to their network so I could pull logs off of a network connected device, just to diagnose an issue they were having. This is of course, unsustainable if you have hardware that’s distributed across multiple locations and want to collect logs more frequently than each time you have physical access to the device.

Reviewer: ```suggestion
I distinctly remember going to out to a customer’s business and connecting to their network so I could pull logs off of a network connected device, just to diagnose an issue they were having. This is of course, unsustainable if you have hardware that’s distributed across multiple locations and want to collect logs more frequently than each time you have physical access to the device.
```

Author: oo that's a lot better

> In that case, you’ll want to build a way to upload logs to a backend somewhere, account for all the storage and privacy-related concerns and regulations, and then a way to download those logs from your computer to analyze them.

Reviewer: ```suggestion
In that case, you’ll want to build a way to upload logs to a backend somewhere, account for all the storage and privacy-related concerns and regulations, and then a way to download those logs to your computer for analysis.
```

### Triggering Collection

> Naively, the easiest way to upload logs is have the user do it. If the device has a UI, it’s as simple as building a button for the user to click when they’re having problems &mdash; think about the "Shake to Report Feedback" or "Submit Feedback" feature on many apps. This way, you can be upfront about the data you’re collecting right, at the time of collection.

Reviewer: ```suggestion
Naively, the easiest way to upload logs is have the user do it. If the device has a UI, it’s as simple as building a button for the user to click when they’re having problems &mdash; think about the "Shake to Report Feedback" or "Submit Feedback" feature on many apps. This way, you can be upfront about the data you’re collecting, right at the time of collection.
```

Sadly, this depends on the user though, and users are notoriously unreliable. If they’re a technical user, they might not even need an UI! — they might be happy using a CLI tool instead. But if they’re non-technical, even asking them to navigate to a certain page might be a huge burden.

> The first alternative that comes to mind is a push system for retrieving logs &mdash; send a notification from the server to the device, and the device will upload the logs when it receives the message. This can take the form of an actual push notification mechanism with a connected web socket, or something as simple as having the device perform a poll every once in a while and collecting the log if the poll tells it to.

Reviewer: This section could use a bit of a re-write imo. 
> The first alternative that comes to mind is a push system for retrieving logs

but then you immediately talk about a polling solution. 

Personally I'd reverse this and say something along the lines of:
The easiest solution is to have your device poll periodically to see if it needs to upload logs. Alternatively, a more advanced system would be to use a push notification system utilizing a websocket or MQTT.

Author: Well a polling solution is basically a push in this context - the developer has to push an request to the device. The mechanism for retrieving that request is either a polling solution or a pushing solution. 

But probably mixing up poll and push in this way won't be as obvious.

Reviewer: gotcha, yeah I see what you mean now. I wasn't thinking about it in that way.

#### Avoid Multiline Logs

Multiline logs are a particularly painful problem for computers to parse. Most system expect logs to all be on a single line, and tracking logs across multiple lines requires retaining state specific to that error message just to parse that error.

```markdown
E bort    : Network error: -55
E bort    :     RSSI: 61
E bort    :     SSID: "Home-Guest"
```

On one line, it’s slightly less visually readable, but vastly more easily machine parsable.

```markdown
E bort    : Network error: -55, RSSI: 61, SSID: "Home-Guest"
```

#### Structured Logs

We can take another page out of server logging best practices [^structuredlogging], and use structured logging. Logs by default are unstructured strings. If we logged more often in a structured format that was easily machine parsable, we could greatly simplify parsing and organizing the data. This is especially useful when we want to eventually parse specific information in the log, otherwise we'd have to write a Regex to parse the log.

```markdown
E bort    : {"Network error": "-55", "RSSI": 61, "SSID": "Home-Guest"}
```

> JSON is the most common format for structured logs, and can be written manually in code fairly quickly. It's also fairly readable compared to the unstructured log, too.

Reviewer: Maybe worth calling out that many platforms have a maximum line length for logs which is easy to blow past in json.

AOSP has some best practices in their [coding guidelines](https://source.android.com/docs/setup/contribute/code-style#log-sparingly) - one of which is ironically "Log Sparingly" and maybe something worth pulling into this article. They recommend log message lenghts up to 80-100 chars and say to avoid lengths of 130-160 **including the tag**.

Author: Good idea about the line length.

Mmm I actually disagree with a lot of general guidance around logging, especially the log sparingly part. It's a totally nonsensical because there's already 200 services logging all the time in AOSP, so they're not even following their own guidance. And I particularly dislike the maximum line length guidance. It's a meaningless recommendation and AOSP exceeds the max all the time as well. The idea should just be, be as concise as you can without losing the information, and log multiple times you if a single line isn't sufficnet. All of this is probably trauma from Py trying to "fix logging".

I actually really like most of the points in the NOTES part, but then it ends with "The golden rule of logging is that your logs may not unnecessarily push other logs out of the buffer, just as others may not push out yours." which is just ??? - the kind of fluff I dislike. 

Honestly you could write a whole separate article on every single idea around improving logging.

#### Skip Log Reading -- Build Visualizations!

The end state of logging is ideally to never read the raw logs ourselves. Once we can empower computers to read all sorts of logs, we can build the tools to parse and then visualize the information, and save ourselves the effort of piecing together all the information in our head.

<p align="center">
  <img width="850" src="{% img_url blog-device-logging/logging-wifi.png %}" alt="Wi-Fi connection status over time" />

> </p>

Reviewer: not specific to this article but I hate how the states are ordered and the colors used. It makes no sense. I hope we can fix that in the future.

## Conclusion

Logging is a fundamental part of our day to day debugging toolset, and it's hard to imagine it ever going away entirely. But we all agree, there is a lot of noise in our logs, there are really no standards for logging, and there's a lot we could probably improve how we log, if we only had the time. Hopefully I've shared some new or helpful ideas around logging that you can put to use today!

<!-- Interrupt Keep START -->
{% include newsletter.html %}

{% include submit-pr.html %}
<!-- Interrupt Keep END -->

{:.no_toc}

## Additional Reading

* [Logging v. instrumentation](https://peter.bourgon.org/blog/2016/02/07/logging-v-instrumentation.html)
* [Tracing: structured logging, but better in every way](https://andydote.co.uk/2023/09/19/tracing-is-better/)

{:.no_toc}

<!-- prettier-ignore-start -->
[^meaningfulmessages]: [Log meaningful messages](https://newrelic.com/blog/best-practices/best-log-management-practices#toc-log-meaningful-messages-that-drive-decisions-)

> [^structuredlogging]: [Structured logging](https://newrelic.com/blog/how-to-relic/structured-logging)

Reviewer: Linking to a semi-competitor, how bold of you.

Author: If you're confident in the product you shouldn't care, I guess.
